File: Java/count_min_sketch/src/main/java/com/alltoolkit/CountMinSketch.java
package com.alltoolkit;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

/**
 * Count-Min Sketch - Probabilistic data structure for frequency estimation.
 * Uses sublinear space for counting item frequencies with controlled error bounds.
 */
public class CountMinSketch<T> {
    private final long[][] table;
    private final int depth;
    private final int width;
    private final long seed;
    private long totalCount;

    public CountMinSketch(int depth, int width, long seed) {
        this.depth = Math.max(1, depth);
        this.width = Math.max(2, width);
        this.seed = seed;
        this.table = new long[this.depth][this.width];
        this.totalCount = 0;
    }

    public CountMinSketch(int depth, int width) {
        this(depth, width, 0xDEADBEEFL);
    }

    /**
     * Create with optimal configuration for given epsilon and delta.
     */
    public static CountMinConfig optimal(double epsilon, double delta) {
        int width = (int) Math.ceil(Math.E / epsilon);
        int depth = (int) Math.ceil(-Math.log(delta));
        return new CountMinConfig(
            Math.max(1, depth),
            Math.max(2, width),
            0xDEADBEEFL
        );
    }

    public static <T> CountMinSketch<T> withRate(double epsilon, double delta) {
        CountMinConfig config = optimal(epsilon, delta);
        return new CountMinSketch<>(config.depth, config.width, config.seed);
    }

    /**
     * Update count for item by delta.
     */
    public void update(T item, long delta) {
        long[] hashes = getHashes(item);
        for (int i = 0; i < depth; i++) {
            int idx = (int) hashes[i];
            table[i][idx] += delta;
        }
        totalCount += delta;
    }

    /**
     * Increment count by 1.
     */
    public void increment(T item) {
        update(item, 1);
    }

    /**
     * Estimate count for item (returns upper bound).
     */
    public long estimate(T item) {
        long[] hashes = getHashes(item);
        long min = table[0][(int) hashes[0]];
        for (int i = 1; i < depth; i++) {
            int idx = (int) hashes[i];
            long val = table[i][idx];
            if (val < min) min = val;
        }
        return min;
    }

    public long totalCount() {
        return totalCount;
    }

    public int[] dimensions() {
        return new int[]{depth, width};
    }

    public void merge(CountMinSketch<T> other) throws DimensionMismatchException {
        if (depth != other.depth || width != other.width) {
            throw new DimensionMismatchException(
                "Cannot merge: depth=" + depth + " vs " + other.depth +
                ", width=" + width + " vs " + other.width
            );
        }
        for (int i = 0; i < depth; i++) {
            for (int j = 0; j < width; j++) {
                table[i][j] += other.table[i][j];
            }
        }
        totalCount += other.totalCount;
    }

    public byte[] toBytes() {
        int size = 32 + depth * width * 8;
        ByteBuffer buf = ByteBuffer.allocate(size).order(ByteOrder.LITTLE_ENDIAN);
        buf.putLong(depth);
        buf.putLong(width);
        buf.putLong(seed);
        buf.putLong(totalCount);
        for (int i = 0; i < depth; i++) {
            for (int j = 0; j < width; j++) {
                buf.putLong(table[i][j]);
            }
        }
        return buf.array();
    }

    public static <T> CountMinSketch<T> fromBytes(byte[] bytes) throws InvalidByteException {
        if (bytes.length < 32) throw new InvalidByteException("Too short: " + bytes.length);
        
        ByteBuffer buf = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
        int d = (int) buf.getLong();
        int w = (int) buf.getLong();
        long s = buf.getLong();
        long tc = buf.getLong();
        
        int expectedLen = 32 + d * w * 8;
        if (bytes.length < expectedLen) throw new InvalidByteException("Invalid length");
        
        CountMinSketch<T> sketch = new CountMinSketch<>(d, w, s);
        sketch.totalCount = tc;
        for (int i = 0; i < d; i++) {
            for (int j = 0; j < w; j++) {
                sketch.table[i][j] = buf.getLong();
            }
        }
        return sketch;
    }

    public void clear() {
        for (int i = 0; i < depth; i++) {
            Arrays.fill(table[i], 0);
        }
        totalCount = 0;
    }

    private long[] getHashes(T item) {
        String str = String.valueOf(item);
        
        long h1 = 0xcbf29ce484222325L;
        for (int i = 0; i < str.length(); i++) {
            h1 = (h1 * 0x100000001b3L) ^ str.charAt(i);
        }
        h1 = (h1 ^ seed) * 0x100000001b3L;
        
        long h2 = seed;
        for (int i = 0; i < str.length(); i++) {
            h2 = (h2 * 0x100000001b3L) ^ str.charAt(i);
        }
        
        long[] hashes = new long[depth];
        for (int i = 0; i < depth; i++) {
            hashes[i] = Math.abs((h1 + i * h2) % width);
        }
        return hashes;
    }

    public static class CountMinConfig {
        public final int depth;
        public final int width;
        public final long seed;

        public CountMinConfig(int depth, int width, long seed) {
            this.depth = depth;
            this.width = width;
            this.seed = seed;
        }
    }

    public static class DimensionMismatchException extends Exception {
        public DimensionMismatchException(String msg) { super(msg); }
    }

    public static class InvalidByteException extends Exception {
        public InvalidByteException(String msg) { super(msg); }
    }

    // Test suite
    public static void runTests() {
        System.out.println("Running Count-Min Sketch tests...");

        // Test 1: Basic increment
        CountMinSketch<String> sketch1 = new CountMinSketch<>(5, 100);
        sketch1.increment("hello");
        sketch1.increment("hello");
        sketch1.increment("world");
        assert sketch1.estimate("hello") >= 2 : "Test 1 failed: hello";
        assert sketch1.estimate("world") >= 1 : "Test 1 failed: world";
        assert sketch1.estimate("missing") == 0 : "Test 1 failed: missing";
        System.out.println("✓ Test 1: Basic increment");

        // Test 2: Update with delta
        CountMinSketch<String> sketch2 = new CountMinSketch<>(5, 100);
        sketch2.update("item", 5);
        assert sketch2.estimate("item") >= 5 : "Test 2 failed";
        System.out.println("✓ Test 2: Update with delta");

        // Test 3: Total count
        CountMinSketch<String> sketch3 = new CountMinSketch<>(5, 100);
        sketch3.increment("a");
        sketch3.update("b", 3);
        sketch3.increment("c");
        assert sketch3.totalCount() == 5 : "Test 3 failed: " + sketch3.totalCount();
        System.out.println("✓ Test 3: Total count");

        // Test 4: Merge
        CountMinSketch<String> sketch4a = new CountMinSketch<>(5, 100);
        CountMinSketch<String> sketch4b = new CountMinSketch<>(5, 100);
        sketch4a.increment("hello");
        sketch4b.increment("world");
        sketch4b.increment("world");
        try {
            sketch4a.merge(sketch4b);
        } catch (DimensionMismatchException e) {
            throw new AssertionError("Merge failed: " + e.getMessage());
        }
        assert sketch4a.estimate("hello") >= 1 : "Test 4 failed: hello";
        assert sketch4a.estimate("world") >= 2 : "Test 4 failed: world";
        System.out.println("✓ Test 4: Merge");

        // Test 5: Clear
        CountMinSketch<String> sketch5 = new CountMinSketch<>(5, 100);
        sketch5.increment("test");
        sketch5.clear();
        assert sketch5.estimate("test") == 0 : "Test 5 failed";
        System.out.println("✓ Test 5: Clear");

        // Test 6: Optimal config
        CountMinConfig config = optimal(0.01, 0.01);
        assert config.depth >= 1 && config.width >= 2 : "Test 6 failed";
        System.out.println("✓ Test 6: Optimal config");

        // Test 7: Serialization
        CountMinSketch<String> sketch7 = withRate("hello");
        sketch7.increment("apple");
        sketch7.increment("banana");
        sketch7.increment("apple");
        byte[] bytes = sketch7.toBytes();
        assert bytes.length > 0 : "Test 7 failed: serialization";
        System.out.println("✓ Test 7: Serialization");

        // Test 8: Dimensions
        CountMinSketch<String> sketch8 = new CountMinSketch<>(7, 200);
        int[] dims = sketch8.dimensions();
        assert dims[0] == 7 && dims[1] == 200 : "Test 8 failed";
        System.out.println("✓ Test 8: Dimensions");

        System.out.println("\n✅ All tests passed!");
    }

    public static <T> CountMinSketch<T> withRate(T item) {
        return withRate(0.01, 0.01);
    }

    public static void main(String[] args) {
        runTests();
    }
}

File: Swift/count_min_sketch/Sources/CountMinSketch.swift
// CountMinSketch.swift
// A probabilistic data structure for frequency estimation

import Foundation

public struct CountMinConfig {
    public let depth: Int
    public let width: Int
    public let seed: UInt64
    
    public init(depth: Int, width: Int, seed: UInt64 = 0xDEADBEEF) {
        self.depth = max(1, depth)
        self.width = max(2, width)
        self.seed = seed
    }
    
    // Create optimal config based on error rate and confidence
    public static func optimal(epsilon: Double, delta: Double) -> CountMinConfig {
        let width = Int Darwin.ceil(M_E / epsilon)
        let depth = Int Darwin.ceil(-Darwin.log(delta))
        return CountMinConfig(depth: max(1, depth), width: max(2, width))
    }
}

public final class CountMinSketch<T: Hashable> {
    private var table: [[UInt64]]
    private let config: CountMinConfig
    private var totalCount: UInt64
    
    public init(config: CountMinConfig) {
        self.config = config
        self.table = Array(repeating: Array(repeating: 0, count: config.width), count: config.depth)
        self.totalCount = 0
    }
    
    public convenience init(depth: Int, width: Int) {
        self.init(config: CountMinConfig(depth: depth, width: width))
    }
    
    public convenience init(epsilon: Double, delta: Double) {
        self.init(config: CountMinConfig(depth: 5, width: 100))
        let optimal = CountMinConfig.optimal(epsilon: epsilon, delta: delta)
        self.table = Array(repeating: Array(repeating: 0, count: optimal.width), count: optimal.depth)
    }
    
    public static func withRate(epsilon: Double, delta: Double) -> CountMinSketch<T> {
        let config = CountMinConfig.optimal(epsilon: epsilon, delta: delta)
        return CountMinSketch<T>(config: config)
    }
    
    // Update count for item by delta
    public func update(_ item: T, delta: UInt64) {
        let hashes = getHashes(item)
        for (i, h) in hashes.enumerated() {
            let idx = Int(h) % config.width
            table[i][idx] += delta
        }
        totalCount += delta
    }
    
    // Increment count by 1
    public func increment(_ item: T) {
        update(item, delta: 1)
    }
    
    // Estimate count (upper bound)
    public func estimate(_ item: T) -> UInt64 {
        let hashes = getHashes(item)
        var minVal: UInt64 = 0
        for (i, h) in hashes.enumerated() {
            let idx = Int(h) % config.width
            let val = table[i][idx]
            if i == 0 || val < minVal {
                minVal = val
            }
        }
        return minVal
    }
    
    public func totalCount() -> UInt64 {
        return totalCount
    }
    
    public func dimensions() -> (depth: Int, width: Int) {
        return (config.depth, config.width)
    }
    
    public func merge(other: CountMinSketch<T>) throws {
        guard config.depth == other.config.depth && config.width == other.config.width else {
            throw CMSError.dimensionMismatch
        }
        for i in 0..<config.depth {
            for j in 0..<config.width {
                table[i][j] += other.table[i][j]
            }
        }
        totalCount += other.totalCount
    }
    
    public func toBytes() -> [UInt8] {
        let size = 32 + config.depth * config.width * 8
        var bytes = [UInt8](repeating: 0, count: size)
        
        var offset = 0
        withUnsafeBytes(of: UInt64(config.depth).littleEndian) { bytes[offset..<offset+8] = $0; offset += 8 }
        withUnsafeBytes(of: UInt64(config.width).littleEndian) { bytes[offset..<offset+8] = $0; offset += 8 }
        withUnsafeBytes(of: config.seed.littleEndian) { bytes[offset..<offset+8] = $0; offset += 8 }
        withUnsafeBytes(of: totalCount.littleEndian) { bytes[offset..<offset+8] = $0; offset += 8 }
        
        for i in 0..<config.depth {
            for j in 0..<config.width {
                withUnsafeBytes(of: table[i][j].littleEndian) { bytes[offset..<offset+8] = $0; offset += 8 }
            }
        }
        return bytes
    }
    
    public static func fromBytes(_ bytes: [UInt8]) throws -> CountMinSketch<String> {
        guard bytes.count >= 32 else { throw CMSError.tooShort }
        
        let depth = Int(UInt64(littleEndian: bytes[0..<8].withUnsafeBytes { $0.load(as: UInt64.self) }))
        let width = Int(UInt64(littleEndian: bytes[8..<16].withUnsafeBytes { $0.load(as: UInt64.self) }))
        let seed = UInt64(littleEndian: bytes[16..<24].withUnsafeBytes { $0.load(as: UInt64.self) })
        let totalCount = UInt64(littleEndian: bytes[24..<32].withUnsafeBytes { $0.load(as: UInt64.self) })
        
        let expectedLen = 32 + depth * width * 8
        guard bytes.count >= expectedLen else { throw CMSError.invalidLength }
        
        let config = CountMinConfig(depth: depth, width: width, seed: seed)
        let sketch = CountMinSketch<String>(config: config)
        
        var offset = 32
        for i in 0..<depth {
            for j in 0..<width {
                let val = UInt64(littleEndian: bytes[offset..<offset+8].withUnsafeBytes { $0.load(as: UInt64.self) })
                sketch.table[i][j] = val
                offset += 8
            }
        }
        return sketch
    }
    
    public func clear() {
        for i in 0..<config.depth {
            for j in 0..<config.width {
                table[i][j] = 0
            }
        }
        totalCount = 0
    }
    
    private func getHashes(_ item: T) -> [UInt64] {
        var h1 = FNVHasher()
        h1.hash(item)
        let sum1 = h1.finalize()
        
        var h2 = FNVHasher()
        h2.hash(config.seed)
        h2.hash(item)
        let sum2 = h2.finalize()
        
        return (0..<config.depth).map { i in sum1 &+ UInt64(i) &* sum2 }
    }
}

public enum CMSError: Error {
    case dimensionMismatch
    case tooShort
    case invalidLength
}

import simd

struct FNVHasher {
    private var state: UInt64 = 0xcbf29ce484222325
    
    mutating func hash<H: Hashable>(_ value: H) {
        var hasher = _FNVHasher(state: &state)
        hasher.combine(value)
    }
    
    mutating func finalize() -> UInt64 {
        return state
    }
}

struct _FNVHasher {
    var state: UnsafeMutablePointer<UInt64>
    
    mutating func combine<T: Hashable>(_ value: T) {
        let bytes = withUnsafeBytes(of: ObjectIdentifier(String(describing: value))) { Array($0) }
        for byte in bytes {
            state.pointee = state.pointee &* 0x100000001b3 &+ UInt64(byte)
        }
    }
}